src/wayland/shell/xdg/mod.rs

```diff
@@ -175,88 +427,150 @@ impl PositionerState {
+            geometry.y += state.anchor_rect.y + state.anchor_rect.height;
+        } else {
+            geometry.y += state.anchor_rect.y + state.anchor_rect.height / 2;
+        }
+
+        if state.anchor_has_edge(xdg_positioner::Anchor::Left) {
+            geometry.x += state.anchor_rect.x;
+        } else if state.anchor_has_edge(xdg_positioner::Anchor::Right) {
+            geometry.x += state.anchor_rect.x + state.anchor_rect.width;
+        } else {
+            geometry.x += state.anchor_rect.x + state.anchor_rect.width / 2;
+        }
+
+        // Adjust for gravity
+        if state.gravity_has_edge(xdg_positioner::Gravity::Top) {
+            geometry.y -= geometry.height;
+        } else if state.gravity_has_edge(xdg_positioner::Gravity::Bottom) {
+            geometry.y -= geometry.height / 2;
+        }
+
+        if state.gravity_has_edge(xdg_positioner::Gravity::Left) {
+            geometry.x -= geometry.width;
+        } else if state.gravity_has_edge(xdg_positioner::Gravity::Right) {
+            geometry.x -= geometry.width / 2;
+        }
+
+        // TODO: Maybe handle constraint_adjustment
+
+        geometry
+    }
 }
 
 /// State of a regular toplevel surface
 #[derive(Debug)]
 pub struct ToplevelState {
-    /// Parent of this surface
-    ///
-    /// If this surface has a parent, it should be hidden
-    /// or displayed, brought up at the same time as it.
-    pub parent: Option<wl_surface::WlSurface>,
-    /// Title of this shell surface
-    pub title: String,
-    /// App id for this shell surface
-    ///
-    /// This identifier can be used to group surface together
-    /// as being several instance of the same app. This can
-    /// also be used as the D-Bus name for the app.
-    pub app_id: String,
     /// Minimum size requested for this surface
     ///
     /// A value of 0 on an axis means this axis is not constrained
-    pub min_size: (i32, i32),
+    pub min_size: Option<(i32, i32)>,
     /// Maximum size requested for this surface
     ///
     /// A value of 0 on an axis means this axis is not constrained
-    pub max_size: (i32, i32),
-}
+    pub max_size: Option<(i32, i32)>,
 
-impl Clone for ToplevelState {
-    fn clone(&self) -> ToplevelState {
-        ToplevelState {
-            parent: self.parent.as_ref().cloned(),
-            title: self.title.clone(),
-            app_id: self.app_id.clone(),
-            min_size: self.min_size,
-            max_size: self.max_size,
-        }
-    }
-}
+    /// The suggested size of the surface
+    pub size: Option<(i32, i32)>,
 
-/// The pending state of a popup surface
-#[derive(Debug)]
-pub struct PopupState {
-    /// Parent of this popup surface
-    pub parent: Option<wl_surface::WlSurface>,
-    /// The positioner specifying how this tooltip should
-    /// be placed relative to its parent.
-    pub positioner: PositionerState,
+    /// The states for this surface
+    // TODO: I would like to use a HashSet here, but the state
```

Comment:
Why a HashSet?

Reply:
With a HashSet there could only be at most one "instance" of each `xdg_toplevel::State` variant. At the moment i check if the state is already in the vec before adding it or otherwise some variant could be inserted multiple times.

Reply:
I have a general intuition that using a hashset for such a small collection (there are rarely more than 3 or 4 states) would induce more overhead than gains, but I admit I may be wrong on that. Do you have a general idea of how it would compare?

If the cost is negligible, I'm ok with using a HashSet.

Reply:
Like discussed on matrix i have added a separate container for handling the states.